### DISCUSSION TITLE
Using `completing-read-multiple` with Vertico/Embark

### DISCUSSION THREAD

**Discussion by alphapapa** [minad/vertico#450] [Q&A]
Hi @minad,

Apologies if this is covered somewhere else, but despite searching this tracker, the Vertico one, the wiki, these discussions, etc, I can't find an answer.

As you might expect, I want to use `completing-read-multiple` with, e.g. Vertico, to select and return multiple candidates.  I've tried using `vertico-insert`, but that overwrites the minibuffer's contents, so I have to manually type each candidate, which defeats the purpose of using completion.

From the searches I've done, it seems that there have been a number of issues and PRs related to this, but it seems like all the explorative approaches were rejected in favor of...well, I can't figure out what.  One issue mentioned Consult, but I can't find anything related to `crm` or `multiple` in Consult.

Am I missing something, or is this still an unsolved problem?

Thanks.

**Comment by oantolin** [comment]
I think there might be something wrong with your configuration. Over here `vertico-insert` (bound to `TAB`) does work. I just tried it again with `describe-face` (which uses crm) to make sure: I can type whatever I want to narrow the candidates and move among them with the arrow keys, when I have the one I want I press `TAB` to insert in it in the minibuffer. Vertico adds it to the minibuffer, overwriting only the portion since the last comma (the default `crm-separator`), keeping the previously selected candidates intact (those before the last comma). Then I can press comma to start the narrowing and selection process over again.

**Comment by alphapapa** [comment]
Hi Omar,

Thanks for your reply.

The problem with that is that it wipes out the partial input, so I have 
to then type it again to show the same list of candidates as before I 
called vertico-insert.  Compare vertico-insert to this:

```el
(defun vertico-insert-or-prepend ()
     "Insert or prepend current candidate in minibuffer.
When `crm-completion-table' is non-nil, assume
`completing-read-multiple' is active, and append the candidate to
the minibuffer input (preserving the trailing input so as to
preserve the existing candidates)."
     (interactive)
     (when (> vertico--total 0)
       (let ((vertico--index (max 0 vertico--index)))
         (if crm-completion-table
             (save-excursion
               (goto-char (+ (point-min) (minibuffer-prompt-width)))
               (insert (vertico--candidate) crm-separator))
           (vertico-insert)))))
```

**Reply by minad** [reply]
(I transferred this discussion to the Vertico repository and edited your comment trying to change the Markdown source block. Unfortunately highlighting still doesn't work.)

> The problem with that is that it wipes out the partial input, so I have
to then type it again to show the same list of candidates as before I
called vertico-insert.  Compare vertico-insert to this:

This is the normal Vertico behavior. Partial input will always be replaced by the selected candidate when pressing TAB.

Your `vertico-insert-or-prepend` command doesn't work correctly, since `crm-separator` is a regular expression. We would need the corresponding separator string. Furthermore the behavior of overwriting partial input is more aligned with normal behavior. However one could check if the partial input matches an actual candidate and in that case automatically insert a separator.

**Reply by alphapapa** [reply]
Well, it works when `crm-separator` is a comma or semicolon.  :)  But it's good to be aware of that limitation.

Regardless of how it's implemented, this seems like sorely missing functionality in Emacs completion (which Helm solved a long time ago, but with its own API, of course), to just be able to type in some input, see some candidates, and choose more than one to return to the calling function, without having the input altered by choosing each candidate (IOW, to have the input and candidates be separate concerns).

I remember seeing discussions on emacs-devel about potential "selecting-read" APIs.  Did those discussions go anywhere, or were they rejected, or did they just stall out?

**Reply by minad** [reply]
> Regardless of how it's implemented, this seems like sorely missing functionality in Emacs completion (which Helm solved a long time ago, but with its own API, of course), to just be able to type in some input, see some candidates, and choose more than one to return to the calling function, without having the input altered by choosing each candidate (IOW, to have the input and candidates be separate concerns).

We've solved this via `completing-read` + `embark-select`. See the comments below, in particular the ones made by @oantolin.

**Comment by oantolin** [comment]
Oh, I see. Wiping out the partial input seems good to me since it seems rare for me to want to add another candidate that matches the same text again. And being forced to erase that partial input would be pretty annoying. I guess the best approach is to have both vertico-input and your command and people can use whatever they want. In fact, binding both is probably best, since I can imagine wanting both in a single crm session

In favor of vertico-input I will say that it very closely matches the crm experience of Emacs's default completion.

**Reply by alphapapa** [reply]
Imagine if `org-ql-find` supported CRM; you could select multiple entries to, e.g. refile.  But as soon as you insert a candidate into the minibuffer, the search query gets wiped out, clearing all the candidates, and you have to type the query again (which might not be just a word or two).

In my case, I'm using my new `listen` package to read tracks from my MPD database.  For example, let's say that I'm searching for an artist named "Bob Smith," and there are matches under a few different names, like "Bob Smith," "Bob Smith Group," "Bob Smith & Charlie Tucker".  I want to add all of those to the minibuffer, separated by the `crm-separator`.  If I use `vertico-insert`, after I add each one, I'll have to type "Bob Smith" again, just to be able to see the other candidates that I was just looking at so I can add them also.

A good comparison is Helm, which supports multiple selection in a very simple way, highlighting each match in the candidates buffer and then acting on them when the user executes the command.  (I realize that Consult may offer similar functionality, which is great, but I'd like my commands that use `completing-read-multiple` to work without having to install extra packages, while also working with ones like Vertico and Orderless and Embark when available.  :)

**Reply by minad** [reply]
> But as soon as you insert a candidate into the minibuffer, the search query gets wiped out, clearing all the candidates, and you have to type the query again (which might not be just a word or two).

Yes, this is a design flaw of CRM. I see your point that it would be nice to preserve the filter input. CRM is generally not a well-designed feature, also since multiple long candidates fill up the prompt line. I recommend to avoid it.

> A good comparison is Helm, which supports multiple selection in a very simple way, highlighting each match in the candidates buffer and then acting on them when the user executes the command.

Multiple selections are supported via Embark. You can select candidates with the `embark-select` command. If `embark-act` is bound to `C-.`, then `embark-select` is available via `C-. SPC`. In my config, I have also bound `embark-select` to `C-SPC`.

**Reply by oantolin** [reply]
Honestly, very few commands seem to use crm. What I use for multiple selections is `embark-select` and `embark-act-all`. You run the command to queue a song, type "Bob Smith" and if you want to queue all the matches you go `C-. A RET` —this uses `embark-act-all` to run the "default action", which for a command that opened a minibuffer is said command. If you want to select only some of the matches, you can mark them with `embark-select` (which you could bind to a key or use `C-. SPC`) and then `embark-act-all` would only work on the selected songs. And of course, embark does not clear the partial input.

Note that this works by looping over the selection and running the command with one argument at a time. I feel it makes it as though every command uses crm even if they don't.

Now, for commands that actually do use crm I guess you might want to see the candidates accumulate in the minibuffer. You could write an `embark-minibuffer-insert` command that is just like `embark-insert` but instead targets the minibuffer (and separates the entries by some configurable separator which you could even check for correctness with `(cl-assert (string-match-p crm-separator separator))`.

**Reply by oantolin** [reply]
Oh, as usual I'm late to the party and @minad already explained `embark-select`. At least I mentioned you're supposed to use it with `embark-act-all`. I like that they bring multiple selection capabilities to all normal Emacs commands even if they were written decades before embark.

**Reply by minad** [reply]
@oantolin Thank you for filling out the missing details! In short, we can do everything we want with `completing-read` + `embark-select` and this combination can completely replace `completing-read-multiple`. 

The advantage here is that all commands can simply continue to use `completing-read` and don't have to loop over multiple candidates. Embark will magically upgrade all single-candidate commands to also support acting on multiple candidates at once.

**Reply by alphapapa** [reply]
> @oantolin Thank you for filling out the missing details! In short, we can do everything we want with `completing-read` + `embark-select` and this combination can completely replace `completing-read-multiple`.
> 
> The advantage here is that all commands can simply continue to use `completing-read` and don't have to loop over multiple candidates. Embark will magically upgrade all single-candidate commands to also support acting on multiple candidates at once.

I wish that were so, but I'm afraid that it may not always be the case.  For example, consider this command in my `listen` package:

```el
(cl-defun listen-queue-add-from-mpd (filenames queue)
  "Add FILENAMES (selected from MPD library) to QUEUE."
  (interactive
   (list (listen-mpd-completing-read :select-tag-p t)
         (listen-queue-complete :allow-new-p t)))
  (require 'listen-mpd)
  (listen-queue-add-files filenames queue))
```
The interactive form first prompts the user to select file(s) with the `listen-mpd-completing-read` function, then to select a queue to which the files are to be added, then it calls `listen-queue-add-files` with the arguments.

If `listen-mpd-completing-read` returns multiple files (as when using CRM), the files are added to the queue, and it's done.

If `embark-select` and `embark-act-all` are used, how will the `queue` argument be preserved across the calls for each candidate?  Won't the user be prompted for it for each file rather than once for all files?

As well, when I try that (using `C-. SPC` to mark each candidate, then `C-. A RET` to call `vertico-exit` on each candidate), I get a long backtrace that ends with:
```
Debugger entered--Lisp error: (no-catch exit nil)
  exit-minibuffer()
  vertico-exit(nil)
  funcall-interactively(vertico-exit nil)
  command-execute(vertico-exit)
```
So it just doesn't seem to work.  This is with Vertico v1.7.

**Reply by oantolin** [reply]
That error is odd! I've tried to reproduce it other commands with no success (and I use this `C-. A RET` pretty often, for example, to attach multiple files to an email with a single call to `mml-attach-file`, so I would have noticed). If it only occurs with `listen`, it might take me a while to reproduce. If you find a simple way to reproduce it, I'd love to get a bug report.

About your other point: you're absolutely right about commands that prompt more than once; you do *normally* get prompted for the second argument each time, but there is a fairly simple way to avoid it. If you write the command to use `completing-read-multiple` for the first argument, to not take the second thing as argument but still prompt for it, and add the command to `embark-multitarget-actions`, then you avoid the repeated prompting. For example, take this command:

```elisp
(defun test-ing (people)
  "Greet all the PEOPLE with an honorific."
  (interactive
   (list (completing-read-multiple "Hi, " '("Ana" "Babs" "Carol" "Dana" "Eve" "Fran" "Greta"))))
  (let ((title (completing-read "Title: " '("Dr." "Ms." "Mrs."))))
    (dolist (person people)
      (insert (format "Hi, %s %s!\n" title person)))))
```

If you use it as-is with the `C-. SPC`, `C-. A RET` method, you will get prompted for each title. But if you also do `(push 'test-ing embark-multitarget-actions)`, then you only get prompted for the title once and it gets used for everyone.

The `embark-multitarget-actions` calling convention calls the command with a single list argument, so you cannot take the `title` parameter as a second argument in the above code, it needs to be prompted for in the body of the command.

Notice that I'm saying that even if a command does uses crm, you, as a user of that command, are not limited to the default crm experience: you can decide instead to use it in a helm-y way by marking with `embark-select` and instead of `RET`, using `C-. A RET`.

**Reply by alphapapa** [reply]
> About your other point: you're absolutely right about commands that prompt more than once; you do _normally_ get prompted for the second argument each time, but there is a fairly simple way to avoid it. If you write the command to use `completing-read-multiple` for the first argument, to not take the second thing as argument but still prompt for it, and add the command to `embark-multitarget-actions`, then you avoid the repeated prompting.

That's great, but, of course, it means that I have to write code specific to `embark`, which, of course, is preferable to avoid (if we're trying to target and enhance the default APIs).

> The embark-multitarget-actions calling convention calls the command with a single list argument, so you cannot take the title parameter as a second argument in the above code, it needs to be prompted for in the body of the command.

So I'd have to rewrite the command to only read one argument interactively?  That's something I'd like to avoid.  I'm trying to follow the Emacs idioms as much as possible, which I think that would violate.

> Notice that I'm saying that even if a command does uses crm, you, as a user of that command, are not limited to the default crm experience: you can decide instead to use it in a helm-y way by marking with embark-select and instead of RET, using C-. A RET.

Now I am a bit confused.  :)  But if the other two points I quoted are true for CRM-using commands as well, then, well, I still don't think the problem is solved very well.  (I don't mean that as a complaint--there's only so much we can do with the default APIs, and you and Daniel, et al have done an amazing job.)

Obviously, what we really need is a `selecting-read`.  @minad @oantolin Do you think there's any chance we could propose such a thing in time for Emacs 30?  (or maybe it could go in ELPA first and then get pulled into core someday)

**Reply by oantolin** [reply]
> That's great, but, of course, it means that I have to write code specific to embark, which, of course, is preferable to avoid (if we're trying to target and enhance the default APIs).

Yes, it would be best to avoid any embark-specific code. Note that the *only* embark-specific code in my example is `(push 'test-ing embark-multitarget-actions)`. The command works just fine for non-Embark users; and Embark users can add that extra bit of configuration (without modifying any code from the package that implements the command) to make it more comfortable to use for `C-. A RET`.

> So I'd have to rewrite the command to only read one argument interactively? That's something I'd like to avoid. I'm trying to follow the Emacs idioms as much as possible, which I think that would violate.

Yes, it sadly is unidiomatic. :( The solution afforded by Embark is not fully general, it only applies to one-argument functions.

>> Notice that I'm saying that even if a command does uses crm, you, as a user of that command, are not limited to the default crm experience: you can decide instead to use it in a helm-y way by marking with embark-select and instead of RET, using C-. A RET.

> Now I am a bit confused. :)

I only meant it for commands that take a single argument which is a list and that prompt for it in an `interactive` form that calls `completing-read-multiple`.

> But if the other two points I quoted are true for CRM-using commands as well, then, well, I still don't think the problem is solved very well. (I don't mean that as a complaint--there's only so much we can do with the default APIs, and you and Daniel, et al have done an amazing job.)

It isn't solved very well. :(

> Obviously, what we really need is a selecting-read.

I've heard people say that and no-one seems to be able to explain how it would differ from completing-read or completing-read-multiple in a way that I can understand. Do you have something specific in mind?

**Reply by alphapapa** [reply]
> I've heard people say that and no-one seems to be able to explain how it would differ from completing-read or completing-read-multiple in a way that I can understand. Do you have something specific in mind?

Well, yes, I suppose: it should work like Helm's multiple selection feature does.  The minibuffer input should not be changed by selecting a candidate; the input and the selected candidates should be independent.

The CRM model with multiple candidate strings separated by a regexp/string just doesn't fit the use case of, e.g. `org-ql-find`, where the input is a query string, not part of a candidate string.  This use case is trivial with Helm, because selecting a candidate doesn't change the input.  IOW, the minibuffer contents are not used as the candidate(s) when the user exits the selection process; the candidate or candidates returned are the one or ones selected with the UI (i.e. the one or ones highlighted in the candidates list).

It seems like a very simple paradigm to me.  It's surely much simpler than incremental completion of strings with common prefixes, isn't it?

**Reply by oantolin** [reply]
I see, that would be good to have. It would be a replacement for crm, then? Sounds good to me.

**Reply by minad** [reply]
@alphapapa Unfortunately you've hit exactly the one use case where a dedicated CRM command can still be useful, a command prompting multiple times. For all other use cases my recommendation is to use `completing-read` and implement all commands such that they process a single candidate. Embark will then handle the rest. I argue that it is better to write commands to handle single candidates, since the command will be simpler, and more convenient to use and more composeable via Embark. Think of it like curried functions in functional programming which operate on single arguments.

However for the use case where you still want CRM there are the these options:

1. Use `completing-read-multiple` and live with its limitations - the candidate separation in the prompt, the inconvenient usage, the difficulty to unselect candidates, ...
2. Call `completing-read` in a loop. Finish the loop if the completion exists with an empty string (on pressing `RET`).
3. Implement a special CRM implementation only for Vertico which replaces the default implementation, and which works like Helm or `embark-select`.

The old `selecting-read` discussion on emacs-devel was orthogonal to `completing-read` and `completing-read-multiple`, and to what we are discussing here. The reason for the `selecting-read` proposal was that there exists some theoretical disagreement about how `completing-read` is supposed to be used. Completion can happen in two modes of operation: Selecting and completing. Selecting means that we select a candidate and return the selected candidate (like Vertico operates). Completing means that the input is incrementally expanded (shell-like, by pressing TAB, like a file path). I don't think adding a separate `selecting-read` API will be an improvement overall, given that we can do both selection and completion already with `completing-read`. The problem here is rather that we need better UI support for CRM (point 3 in my list). Helm already provides a better CRM UI, but Vertico does not. We could also add such an improved CRM UI to Vertico. This means no API change is needed on the side of Emacs.
